sigstore/_verify/policy.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2022 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+APIs for describing identity verification "policies", which describe how the identities
+passed into an individual verification step are verified.
+"""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+
+try:
+    from typing import Protocol
+except ImportError:
+    # TODO(ww): Remove when our minimum Python is 3.8.
+    from typing_extensions import Protocol  # type: ignore[assignment]
+
+from cryptography.x509 import (
+    Certificate,
+    ExtensionNotFound,
+    ObjectIdentifier,
+    OtherName,
+    RFC822Name,
+    SubjectAlternativeName,
+    UniformResourceIdentifier,
+)
+from pydantic import BaseModel
+
+# From: https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md
+_OIDC_ISSUER_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.1")
+_OIDC_GITHUB_WORKFLOW_TRIGGER_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.2")
+_OIDC_GITHUB_WORKFLOW_SHA_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.3")
+_OIDC_GITHUB_WORKFLOW_NAME_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.4")
+_OIDC_GITHUB_WORKFLOW_REPOSITORY_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.5")
+_OIDC_GITHUB_WORKFLOW_REF_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.6")
+_OTHERNAME_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.7")
+
+
+class VerificationResult(BaseModel):
+    success: bool
+
+    def __bool__(self) -> bool:
+        return self.success
+
+
+class VerificationSuccess(VerificationResult):
+    success: bool = True
+
+
+class VerificationFailure(VerificationResult):
+    success: bool = False
+    reason: str
+
+
+class VerificationPolicy(Protocol):
+    @abstractmethod
+    def verify(self, cert: Certificate) -> VerificationResult:
+        raise NotImplementedError
+
+
+class AnyOf:
```

Comment:
Should these inherit from `VerificationPolicy`?

Reply:
I'm trying out a substructural pattern here :slightly_smiling_face: -- `VerificationPolicy` is a protocol type, so when MyPy sees it in a signature it understands that any type that implements the `VerificationPolicy` interface is valid, rather than just child classes.

(I tried doing this originally with ABCs, but switched to protocols since they're more modern. But I can switch it back if you think it'll help improve comprehension here!)

Reply:
Neat! No need to switch back, I just hadn't seen this before.